src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs

```diff
@@ -385,136 +385,48 @@ public void Encode(byte value)
-                                    WriteByte(byte5);
-                                    WriteByte(byte6);
-                                    WriteByte((byte)value);
-                                    return;
-                                }
-
-                                WriteByte(132);
-                                WriteByte(byte4);
-                                WriteByte(byte5);
-                                WriteByte(byte6);
-                                WriteByte((byte)value);
-                                return;
-                            }
-
-                            WriteByte(133);
-                            WriteByte(byte3);
-                            WriteByte(byte4);
-                            WriteByte(byte5);
-                            WriteByte(byte6);
-                            WriteByte((byte)value);
-                            return;
-                        }
-
-                        WriteByte(134);
-                        WriteByte(byte2);
-                        WriteByte(byte3);
-                        WriteByte(byte4);
-                        WriteByte(byte5);
-                        WriteByte(byte6);
-                        WriteByte((byte)value);
-                        return;
-                    }
-
-                    WriteByte(135);
-                    WriteByte(byte1);
-                    WriteByte(byte2);
-                    WriteByte(byte3);
-                    WriteByte(byte4);
-                    WriteByte(byte5);
-                    WriteByte(byte6);
-                    WriteByte((byte)value);
-                    return;
-                }
+            value = BinaryPrimitives.ReverseEndianness(value);
+            Span<byte> valueSpan = MemoryMarshal.CreateSpan(ref Unsafe.As<ulong, byte>(ref value), sizeof(ulong));
+            // Ok to stackalloc even if we don't use with SkipLocalsInit
+            Span<byte> output = stackalloc byte[1 + sizeof(ulong)];
 
-                WriteByte(136);
-                WriteByte(byte0);
-                WriteByte(byte1);
-                WriteByte(byte2);
-                WriteByte(byte3);
-                WriteByte(byte4);
-                WriteByte(byte5);
-                WriteByte(byte6);
-                WriteByte((byte)value);
-                return;
+            byte prefix = (byte)(0x80 + valueLength);
+            if (leadingZeroBytes > 0)
+            {
+                // Reuse space in valueSpan for prefix rather than copying
+                valueSpan[leadingZeroBytes - 1] = prefix;
+                output = valueSpan.Slice(leadingZeroBytes - 1, 1 + valueLength);
+            }
+            else
+            {
+                // Build final output: prefix + value bytes
+                output[0] = prefix;
+                valueSpan.Slice(leadingZeroBytes, valueLength).CopyTo(output.Slice(1));
+                output = output.Slice(0, 1 + valueLength);
```

Comment:
wouldn't it be better to not using `output` at all and have one Write(byte) for prefix and then one Write(valuespan.Slice)?

Reply:
Issue is more that they are virtual calls and have very different behaviour if NettyStream or KeccakStream; so faster just to do the work on the local spans

Reply:
2 non inlinable virtual calls; each calling 2 more non-inlinable interface calls (for `NettyRlpStream`); potentially reallocating and moving memory, span is just a pointer bump due to fixed small size and `SkipLocalsInit`